src/compiler/checker.ts

```diff
@@ -8579,7 +8586,8 @@ namespace ts {
 
         /** Return the inferred type for a binding element */
         function getTypeForBindingElement(declaration: BindingElement): Type | undefined {
-            const parentType = getTypeForBindingElementParent(declaration.parent.parent);
+            const checkMode = declaration.dotDotDotToken && declaration.parent.kind !== SyntaxKind.ArrayBindingPattern ? CheckMode.ObjectRestBindingElement : CheckMode.Normal;

```

Comment:
I assumed here we only want this special rest behavior for rest elements in *object* binding patterns, but maybe we also want it for array rest elements too.